src/components/RouteLayout.jsx

```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import Switch from 'react-router-dom/Switch';
+import Route from 'react-router-dom/Route';
+
+const RouteWithSubRoutes = route => {
+	if (!route.component) {
+		throw new Error(`route for ${route.path} must have a 'component' property`);
+	}
+	if (route.render || route.children) {
+		console.warn('route.render and route.children function not supported');
+	}
+	return (
+		<Route
+			path={route.path}
+			exact={route.exact || false}
+			strict={route.strict || false}
+			render={props => (
+				<route.component {...props}>
+					{route.routes && <SwitchRoutes routes={route.routes} />}
+				</route.component>
+			)}
+		/>
+	);
+};
+
+const SwitchRoutes = ({ routes }) => (
+	<Switch>
+		{routes.map((route, i) =>
+			<RouteWithSubRoutes key={i} {...route} />
```

Comment:
There are definitely ways that `RouteWithSubRoutes` could be extended to support different route styles (e.g. using [`render`](https://reacttraining.com/react-router/#route.render) or [`children`](https://reacttraining.com/react-router/#route.children) instead of `component`) - but the current setup essentially supports a `routes` definition that closely parallels the routes definition for RRv3, which keeps it simpler for apps to upgrade to v4 in the short term.

Over time, we should be able to add flexibility, but I'm reluctant to do that now before we know how we're going to manage async routes and how the query system will adapt to different route compositions.